### PR TITLE
Chapter 12

### PR DESCRIPTION
**Pull request by heimrih**
Chapter 12 - Forget Password

### CODE REVIEW

**Comment by heimrih** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update]
+  before_action :valid_user, only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: image

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  def new; end
+
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove ()

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    else
+      flash.now[:danger] = t "forgot.fail"
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: .blank?

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      @user.errors.add :password, t("forgot.reset.error")
+      render :edit
+    elsif @user.update(user_params)
+      log_in @user
+      @user.update_attribute(:reset_digest, nil)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove ()

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      flash[:success] = t "forgot.reset.success"
+      redirect_to @user
+    else
+      render :edit
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:33]: add flash for this case

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:password, :password_confirmation)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove ()

File: app/controllers/password_resets_controller.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def get_user
+    @user = User.find_by(email: params[:email])
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove ()

File: app/controllers/password_resets_controller.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def valid_user
+    unless (@user && @user.activated? && @user.authenticated?(:reset, params[:id]))
+    redirect_to root_path
+    end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
return if @user && @user.authenticated? :reset, params[:id]
    redirect_to root_path
```and add flash

File: app/controllers/password_resets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
+      flash[:danger] = t "forgot.reset.sessionexpire"
+    redirect_to new_password_reset_url
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
      redirect_to new_password_reset_url
```

File: app/helpers/password_resets_helper.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/password_resets_helper.rb]: remove file that never used

File: app/mailers/user_mailer.rb [hunk 11]
```diff
@@ -3,4 +3,9 @@ def account_activation user
     @user = user
     mail to: user.email, subject: t("user.activation.activation")
   end
+
+  def password_reset(user)
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: remove ()

File: app/models/user.rb [hunk 12]
```diff
@@ -57,6 +57,19 @@ def send_activation_email
     UserMailer.account_activation(self).deliver_now
   end
 
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_columns(reset_digest: FILL_IN, reset_sent_at: FILL_IN)
```
Comment (saobv-2223) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 13]
```diff
@@ -57,6 +57,19 @@ def send_activation_email
...
+  end
+
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
```
Comment (saobv-2223) [inline app/models/user.rb]: Please, move 2 to settings file

File: app/views/password_resets/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,17 @@
+<% provide :title, t("forgot.reset.title") %>
+<title><%= t "forgot.reset.title" %></title>
+<h1><%= t "forgot.reset.title" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: password_reset_path(params[:id])) do |f| %>
```
Comment (saobv-2223) [inline app/views/password_resets/edit.html.erb]: remove ()

File: app/views/password_resets/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,13 @@
+<% provide :title, t("forgot.title") %>
+<title><%= t "forgot.title" %></title>
+<h1><%= t "forgot.title" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:password_reset, url: password_resets_path) do |f| %>
```
Comment (saobv-2223) [inline app/views/password_resets/new.html.erb]: remove ()

File: config/routes.rb [hunk 16]
```diff
@@ -8,8 +8,12 @@
     get "/login", to: "sessions#new", as: :login
     post "/login", to: "sessions#create"
     delete "/logout", to: "sessions#destroy", as: :logout
+    get "password_resets/new"
+    get "password_resets/edit"
     resources :users
     resources :account_activations, only: :edit
+
+    resources :password_resets, only: [:new, :create, :edit, :update]
```
Comment (saobv-2223) [inline config/routes.rb]: image

**Comment by heimrih** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+  def get_user
+    @user = User.find_by email: params[:email]
+    return if @user
+
+    flash[:warning] = t "user.new.notfound"
+    redirect_to root_path
+  end
+
+  def valid_user
```
Comment (ngocvt-0484) [inline app/controllers/password_resets_controller.rb:51]:
```suggestion
  def valid_user
  return if @user && @user.authenticated? :reset, params[:id]
  flash[:warning] = t "user.new.notfound"
  redirect_to root_path
end
```

File: app/controllers/password_resets_controller.rb [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+    return if @user && @user.authenticated? :reset, params[:id]
+      redirect_to root_path
+      else
+      flash[:warning] = t "user.new.notfound"
+      redirect_to root_path
+    end
+  end
+
+  def check_expiration
```
Comment (ngocvt-0484) [inline app/controllers/password_resets_controller.rb:57]:
```suggestion
  def check_expiration
  return unless @user.password_reset_expired?
  flash[:danger] = t "forgot.reset.sessionexpire"
  redirect_to new_password_reset_url
end
```

**Comment by heimrih** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    if params[:user][:password].blank?
+      @user.errors.add :password, t("forgot.reset.error")
+      render :edit
+    elsif @user.update user_params
+      log_in @user
+      @user.update_attribute :reset_digest, nil
+      flash[:success] = t "forgot.reset.success"
+      redirect_to @user
+    else
+      flash[:warning] = t "forgot.reset.failed"
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
      flash.now[:warning] = t "forgot.reset.failed"
```

**Comment by heimrih** [comment]
ready

**Comment by heimrih** [comment]
ready

**Comment by thaodtd-0665** [comment]
ready

**Review by chungpv-1008 — COMMENTED** [review]
merge

**Pull request merged**